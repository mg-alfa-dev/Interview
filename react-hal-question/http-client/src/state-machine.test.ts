import _ from 'lodash';
import defaultState from './default-state';
import stateMachine from './state-machine';

const { __setNewBranchId } = require('./helpers') as typeof import('./__mocks__/helpers');

jest.mock('./helpers');
const [repo1Id, repo2Id] = Object.values(defaultState.repositories).map(r => r.id);
const [branch1id, branch2id] = Object.values(defaultState.branches).map(b => b.id);

beforeEach(() => {
  stateMachine.reset();
});

afterEach(() => {
  __setNewBranchId('foo');
});

test('call reset() resets to default state', () => {
  delete stateMachine.repositories.b5ob3r;
  delete stateMachine.branches.rvpylf;

  stateMachine.reset();

  expect(stateMachine).toMatchObject(defaultState);
});

test('getting the repository list returns the repositories document as expected', async () => {
  expect(await stateMachine.getRepositoryList()).toMatchSnapshot();
});

test('getting a repository returns the repository document as expected', async () => {
  expect(await stateMachine.getRepository(repo1Id)).toMatchSnapshot();
});

test('getting a repository that does not exist throws an error', async () => {
  await expect(stateMachine.getRepository('foo')).rejects.toMatchInlineSnapshot(`[Error: Repository with id 'foo' does not exist.]`);
});

test('getting branch details returns the branch document as expected', async () => {
  expect(await stateMachine.getBranchDetails(branch1id, repo1Id)).toMatchSnapshot();
});

test('getting a branch that does not exist on a given repo throws an error', async () => {
  await expect(stateMachine.getBranchDetails(branch2id, repo2Id)).rejects.toMatchInlineSnapshot(`[Error: Branch with id 'k0dceo' not found in repository with id '5mcf7i'.]`);
});

test('adding a branch acts as expected', async () => {
  expect(await stateMachine.createBranch(repo1Id, 'bar')).toMatchSnapshot({ item: { dateCreated: expect.any(String) } }, 'response data');
  expect(stateMachine.branches[JSON.stringify([repo1Id, 'foo'])]).toMatchSnapshot(
    {
      dateCreated: expect.any(Date),
    },
    'added branch data'
  );
  expect(await stateMachine.getRepository(repo1Id)).toMatchSnapshot('repository response data');
  expect(await stateMachine.getBranchDetails('foo', repo1Id)).toMatchSnapshot(
    {
      item: {
        dateCreated: expect.any(String),
      },
    },
    'branch details response data'
  );
});

test('adding a branch to a readonly repository throws an error', async () => {
  await expect(stateMachine.createBranch(repo2Id, 'bar')).rejects.toMatchInlineSnapshot(
    `[Error: You do not have permissions to create a branch on repository 'Walk unscheduled (read-only)']`
  );
});

test('adding a branch to a non-existent repository throws an error', async () => {
  await expect(stateMachine.createBranch('lol wut?', 'bar')).rejects.toMatchInlineSnapshot(`[Error: Repository with id 'lol wut?' does not exist.]`);
});
