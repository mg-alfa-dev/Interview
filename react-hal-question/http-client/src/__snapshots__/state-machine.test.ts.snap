// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adding a branch acts as expected: added branch data 1`] = `
Object {
  "dateCreated": Any<Date>,
  "details": "Lorem ipsum and all that nonsense",
  "id": "foo",
  "links": Array [
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=foo",
      "rel": "self",
    },
  ],
  "name": "bar",
  "repositoryId": "e2mh4q",
}
`;

exports[`adding a branch acts as expected: branch details response data 1`] = `
Object {
  "contentType": "application/vnd.Branch",
  "item": Object {
    "dateCreated": Any<String>,
    "details": "Lorem ipsum and all that nonsense",
    "id": "foo",
    "links": Array [
      Object {
        "contentType": "application/vnd.Branch",
        "href": "/repository/e2mh4q/fetchBranchDetails?id=foo",
        "rel": "self",
      },
    ],
    "name": "bar",
    "repositoryId": "e2mh4q",
  },
  "links": Array [
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=foo",
      "rel": "self",
    },
  ],
}
`;

exports[`adding a branch acts as expected: repository response data 1`] = `
Object {
  "contentType": "application/vnd.Repository",
  "item": Object {
    "branchSummaries": Object {
      "b5ob3r": Object {
        "id": "b5ob3r",
        "name": "Branch 1",
        "repositoryId": "e2mh4q",
      },
      "foo": Object {
        "id": "foo",
        "name": "bar",
        "repositoryId": "e2mh4q",
      },
      "k0dceo": Object {
        "id": "k0dceo",
        "name": "Branch 2",
        "repositoryId": "e2mh4q",
      },
    },
    "id": "e2mh4q",
    "name": "Run schedule",
  },
  "links": Array [
    Object {
      "href": "/repository/e2mh4q",
      "rel": "self",
    },
    Object {
      "href": "/repository/e2mh4q/add-branch?name={{name}}",
      "rel": "AddBranch",
    },
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=b5ob3r",
      "rel": "FetchDetails-b5ob3r",
    },
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=k0dceo",
      "rel": "FetchDetails-k0dceo",
    },
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=foo",
      "rel": "FetchDetails-foo",
    },
  ],
}
`;

exports[`adding a branch acts as expected: response data 1`] = `
Object {
  "contentType": "application/vnd.Branch",
  "item": Object {
    "dateCreated": Any<String>,
    "details": "Lorem ipsum and all that nonsense",
    "id": "foo",
    "links": Array [
      Object {
        "contentType": "application/vnd.Branch",
        "href": "/repository/e2mh4q/fetchBranchDetails?id=foo",
        "rel": "self",
      },
    ],
    "name": "bar",
    "repositoryId": "e2mh4q",
  },
  "links": Array [
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=foo",
      "rel": "self",
    },
  ],
}
`;

exports[`getting a repository returns the repository document as expected 1`] = `
Object {
  "contentType": "application/vnd.Repository",
  "item": Object {
    "branchSummaries": Object {
      "b5ob3r": Object {
        "id": "b5ob3r",
        "name": "Branch 1",
        "repositoryId": "e2mh4q",
      },
      "k0dceo": Object {
        "id": "k0dceo",
        "name": "Branch 2",
        "repositoryId": "e2mh4q",
      },
    },
    "id": "e2mh4q",
    "name": "Run schedule",
  },
  "links": Array [
    Object {
      "href": "/repository/e2mh4q",
      "rel": "self",
    },
    Object {
      "href": "/repository/e2mh4q/add-branch?name={{name}}",
      "rel": "AddBranch",
    },
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=b5ob3r",
      "rel": "FetchDetails-b5ob3r",
    },
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=k0dceo",
      "rel": "FetchDetails-k0dceo",
    },
  ],
}
`;

exports[`getting branch details returns the branch document as expected 1`] = `
Object {
  "contentType": "application/vnd.Branch",
  "item": Object {
    "dateCreated": "Fri, 01 Oct 2021 20:42:46 GMT",
    "details": "I am Branch 1.  Enjoy.",
    "id": "b5ob3r",
    "links": Array [
      Object {
        "href": "/e2mh4q/fetchBranchDetails?id=b5ob3r",
        "rel": "fetchDetails-b5ob3r",
      },
    ],
    "name": "Branch 1",
    "repositoryId": "e2mh4q",
  },
  "links": Array [
    Object {
      "contentType": "application/vnd.Branch",
      "href": "/repository/e2mh4q/fetchBranchDetails?id=b5ob3r",
      "rel": "self",
    },
  ],
}
`;

exports[`getting the repository list returns the repositories document as expected 1`] = `
Object {
  "contentType": "application/vnd.RepositoryList",
  "item": Array [
    Object {
      "contentType": "application/vnd.Repository",
      "item": Object {
        "branchSummaries": Object {
          "b5ob3r": Object {
            "id": "b5ob3r",
            "name": "Branch 1",
            "repositoryId": "e2mh4q",
          },
          "k0dceo": Object {
            "id": "k0dceo",
            "name": "Branch 2",
            "repositoryId": "e2mh4q",
          },
        },
        "id": "e2mh4q",
        "name": "Run schedule",
      },
      "links": Array [
        Object {
          "href": "/repository/e2mh4q",
          "rel": "self",
        },
        Object {
          "href": "/repository/e2mh4q/add-branch?name={{name}}",
          "rel": "AddBranch",
        },
        Object {
          "contentType": "application/vnd.Branch",
          "href": "/repository/e2mh4q/fetchBranchDetails?id=b5ob3r",
          "rel": "FetchDetails-b5ob3r",
        },
        Object {
          "contentType": "application/vnd.Branch",
          "href": "/repository/e2mh4q/fetchBranchDetails?id=k0dceo",
          "rel": "FetchDetails-k0dceo",
        },
      ],
    },
    Object {
      "contentType": "application/vnd.Repository",
      "item": Object {
        "branchSummaries": Object {
          "b5ob3r": Object {
            "id": "b5ob3r",
            "name": "Branch 3",
            "repositoryId": "5mcf7i",
          },
          "rvpylf": Object {
            "id": "rvpylf",
            "name": "Branch 4",
            "repositoryId": "5mcf7i",
          },
        },
        "id": "5mcf7i",
        "name": "Walk unscheduled (read-only)",
      },
      "links": Array [
        Object {
          "href": "/repository/5mcf7i",
          "rel": "self",
        },
        Object {
          "contentType": "application/vnd.Branch",
          "href": "/repository/5mcf7i/fetchBranchDetails?id=b5ob3r",
          "rel": "FetchDetails-b5ob3r",
        },
        Object {
          "contentType": "application/vnd.Branch",
          "href": "/repository/5mcf7i/fetchBranchDetails?id=rvpylf",
          "rel": "FetchDetails-rvpylf",
        },
      ],
    },
  ],
  "links": Array [
    Object {
      "href": "/repositoryList",
      "rel": "self",
    },
  ],
}
`;
